src/dataflow/src/render/upsert.rs

```diff
@@ -123,149 +122,39 @@ where
-                                        // retract old value
-                                        session.give((old_value, cap.time().clone(), -1));
-                                    }
-                                    if let Some(new_value) = new_value {
-                                        // give new value
-                                        session.give((new_value, cap.time().clone(), 1));
-                                    }
-                                }
-                                None => {}
-                                Some(Err(err)) => {
-                                    // This can never be retracted! But at least it's better to put the source in a
-                                    // permanently errored state than to keep on trucking with wrong results.
-                                    session.give((Err(err), cap.time().clone(), 1));
-                                }
-                            }
-                        }
-                    } else {
-                        // because this is a BTreeMap, the rest of the times in
-                        // the map will be greater than this time. So if the
-                        // input_frontier is less than or equal to this time,
-                        // it will be less than the times in the rest of the map
-                        break;
-                    }
-                }
-                // Discard entries, capabilities for complete times.
-                for time in removed_times {
-                    to_send.remove(&time);
+    let upsert_output = stream.upsert(
+        |decode_result| &decode_result.key,
+        |decode_result| {
+            (
+                decode_result.key,
+                decode_result.value,
+                decode_result.position.expect("missing Kafka offset"),
+            )
+        },
+        move |key: &Result<Row, _>, value: Result<Row, _>| {
+            match key {
+                Ok(key) => value
+                    .and_then(|value| {
+                        let mut datums = Vec::with_capacity(source_arity);
+                        datums.extend(key.iter());
+                        datums.extend(value.iter());
+                        evaluate(&datums, &predicates, &position_or, &mut row_packer)
+                            .map_err(DataflowError::from)
+                    })
+                    .transpose(),
+                Err(err) => {
+                    // This can never be retracted! But at least it's better to put the source in a
+                    // permanently errored state than to keep on trucking with wrong results.
+                    Some(Err(err.clone()))
                 }
             }
         },
+        as_of_frontier,
     );
 
-    let (mut oks, mut errs) = result_stream.ok_err(|(data, time, diff)| match data {
+    let (mut oks, mut errs) = upsert_output.ok_err(|((_key, data), time, diff)| match data {
```

Comment:
The stream now returns `(key, data)` instead of just `data`. Why is that?

Is there any memory impact from the stream having rows containing `(key, data)` instead of just `data`?

Reply:
The upsert operator was already storing the key as well, it's just that we didn't emit it previously. When observing memory usage on the `avro-upsert` benchmark, it looks like there is no difference between the old and new operator.

But: I could remove the key from the output again, this is partially an artifact of how I was developing it and using it in the `kafka_upsert.rs` example.